sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,539 @@
+feature.
+
+## Background
+Sometimes users want to link device code dynamically at run time. One possible
+use case for such linkage - providing device functions via shared libraries.
+The example below shows how device function `LibDeviceFunc` can be dynamically
+linked to a SYCL app:
+```
+// app.cpp
+SYCL_EXTERNAL int LibDeviceFunc(int i);
+class KernelName;
+/* ... */
+Q.submit([&](cl::sycl::handler &CGH) {
+CGH.parallel_for<KernelName>(/* ... */ [=](sycl::item i) {
+  out[i] = LibDeviceFunc(i);
+}); /* ... */
+std::cout << out[i] << “ “;
+
+// lib.cpp
+int SYCL_EXTERNAL LibDeviceFunc(int i) {
+  return i * 2;
+}
+
+// Commands
+clang++ -fsycl lib.cpp -shared -o helpers.so
+clang++ -fsycl app.cpp -lhelpers -o a.out
+./a.out
+Output: 0 2 4 6…
+
+```
+It is possible to manually create `sycl::program` in both app and shared
+library, then use `link` SYCL API to get a single program and launch kernels
+using it. But it is not user-friendly and it is very different from regular
+C/C++ workflow.
+
+Another possible scenario - use functions defined in a pre-compiled device image
+provided by the user. Example:
+```
+// a.cpp
+SYCL_EXTERNAL void LibDeviceFunc();
+...
+Q.submit([&](cl::sycl::handler &CGH) {
+CGH.parallel_for([]() { LibDeviceFunc(); });
+});
+
+// b.cpp
+/*no SYCL_EXTERNAL*/ void LibDeviceFunc() { ... }
+```
+We have a `SYCL_EXTERNAL` function `foo` called from a kernel, but the
+application defined only host version of this function. Then user adds device
+image with definition of `foo` to the fat object via special compiler option
+(like `-fsycl-add-targets`).
+
+The main purpose of this feature is to provide a mechanism which allows to
+link device code dynamically at runtime, such as in the scenarios above.
+
+## Requirements:
+User's device code that consists of some device API (`SYCL_EXTERNAL` functions),
+is compiled into some form and it is not linked statically with device code of
+application. It can be a shared library that contains some device code or a
+separate device image supplied with properties attached. This code is linked
+dynamically at run time with device code of a user's application in order to
+resolve dependencies.
+For this combination the following statements must be true:
+
+The presented dynamic device code linkage mechanism must:
+
+- Allow to represent the actual dynamically linked code as a device binary image
+  which can be:
+  - Embedded into a host shared object by standard SYCL compiler driver
+    invocation
+  - Embedded into a host binary or shared object using manual invocations of
+    SYCL tools such as `clang-offload-wrapper` and linker
+  - Loaded into memory via special API
+- Must not assume the actual format of the device code - e.g. that it is SPIR-V
+  or native device binary
+- Provide automatic runtime resolution of `SYCL_EXTERNAL` function references
+  within the SYCL app to their definitions (if found) within any suitable
```

Comment:
> any suitable dynamically linked device binary image

Is it linked to the **host** binary or where? Please clarify where the search for "device code" is looking into

Reply:
> Is it linked to the host binary or where?

It is linked to other device binary images.

> Please clarify where the search for "device code" is looking into

I've tried to clarify it here:
https://github.com/intel/llvm/pull/3210/files/9f2b787fdf609f1d81e48eb397262a773e4ed094#r618274713

Reply:
Sorry, I don't see clarifications in the link you provided.

Reply:
Sorry for inconvenience. I meant the comment where I mentioned that there is an internal storage for device images inside `ProgramManager` runtime class. It stores all device images available during application runtime. Any search for "device code" will happen inside this storage and it will be search for some device image. If this is not what you were asking for, could you please clarify your question then?